File: src/main/java/org/geepawhill/contentment/position/Placement.java
package org.geepawhill.contentment.position;

import org.geepawhill.contentment.core.NodeSource;
import org.geepawhill.contentment.geometry.PointPair;
import org.geepawhill.contentment.utility.JfxUtility;

import javafx.geometry.*;
import javafx.scene.Node;

public class Placement implements Position
{
	private PointPair area;
	private HPos horizontal;
	private NodeSource above;

	public Placement(PointPair area, NodeSource above, HPos horizontal)
	{
		this.area = area;
		this.above = above;
		this.horizontal = horizontal;
	}

	@Override
	public void position(Node node, PointPair dimensions)
	{
		PointPair remainder = area;
		if(above!= NodeSource.Companion.getNONE())
		{
			double newY = new PointPair(above.get().getBoundsInParent()).south().getY();
			remainder = new PointPair(remainder.getFrom().getX(),newY, remainder.getTo().getX(), remainder.getTo().getY());
		}
		switch(horizontal)
		{
		case LEFT:
			node.setTranslateX(remainder.west().getX());
			break;
		case RIGHT:
			node.setTranslateX(remainder.east().getX() -dimensions.width());
			break;
		case CENTER:
			node.setTranslateX(remainder.centerX()-dimensions.width()/2);
		}
		
		node.setTranslateY(remainder.north().getY());
		JfxUtility.setVerticalAlignment(node,VPos.TOP);
	}
}

File: src/main/kotlin/org/geepawhill/contentment/position/Placement.kt
package org.geepawhill.contentment.position

import org.geepawhill.contentment.core.NodeSource
import org.geepawhill.contentment.geometry.PointPair
import org.geepawhill.contentment.utility.JfxUtility

import javafx.geometry.*
import javafx.scene.Node

class Placement(private val area: PointPair, private val above: NodeSource, private val horizontal: HPos) : Position {

    override fun position(node: Node, dimensions: PointPair) {
        var remainder = area
        if (above !== NodeSource.NONE) {
            val newY = PointPair(above.get().boundsInParent).south().y
            remainder = PointPair(remainder.from.x, newY, remainder.to.x, remainder.to.y)
        }
        when (horizontal) {
            HPos.LEFT -> node.translateX = remainder.west().x
            HPos.RIGHT -> node.translateX = remainder.east().x - dimensions.width()
            HPos.CENTER -> node.translateX = remainder.centerX() - dimensions.width() / 2
        }

        node.translateY = remainder.north().y
        JfxUtility.setVerticalAlignment(node, VPos.TOP)
    }
}
